addons/controllers/clusterbootstrap_controller.go

```diff
@@ -278,14 +320,270 @@ func (r *ClusterBootstrapReconciler) createOrPatchclusterBootstrapFromTemplate(c
+	pkgi := &kapppkgiv1alpha1.PackageInstall{
+		ObjectMeta: metav1.ObjectMeta{
+			// The legacy addons controller uses <cluster name>-<addon name> convention for naming the PackageInstall CR.
+			// https://github.com/vmware-tanzu/tanzu-framework/blob/main/addons/controllers/package_reconciler.go#L195.
+			// util.GeneratePackageInstallName() follows the same pattern.
+			Name: util.GeneratePackageInstallName(cluster.Name, util.GetPackageShortName(clusterBootstrap.Spec.Kapp.RefName)),
+			// kapp-controller PackageInstall CR is installed under the same namespace as tanzuClusterBootstrap. The namespace
+			// is also the same as where the cluster belongs.
+			Namespace: clusterBootstrap.Namespace,
+		},
+	}
+
+	pkgiMutateFn := func() error {
+		// TODO: Followup on the following fields, only add them if needed.
+		// https://github.com/vmware-tanzu/tanzu-framework/issues/1677
+		// if ipkg.ObjectMeta.Annotations == nil {
+		//	 ipkg.ObjectMeta.Annotations = make(map[string]string)
+		// }
+		// ipkg.ObjectMeta.Annotations[addontypes.YttMarkerAnnotation] = ""
+		// ipkg.Spec.SyncPeriod = &metav1.Duration{Duration: r.Config.AppSyncPeriod}
+		// ipkg.Spec.ServiceAccountName = r.Config.AddonServiceAccount
+		pkgi.Spec.PackageRef = &kapppkgiv1alpha1.PackageRef{
+			RefName: clusterBootstrap.Spec.Kapp.RefName,
+			VersionSelection: &versions.VersionSelectionSemver{
+				Prereleases: &versions.VersionSelectionSemverPrereleases{},
+			},
+		}
+		pkgi.Spec.Values = []kapppkgiv1alpha1.PackageInstallValues{
+			{SecretRef: &kapppkgiv1alpha1.PackageInstallValuesSecretRef{
+				// The secret name could also be fetched from kappConfig.Spec.Status.SecretRef. However, the naming convention
+				// of that field is the same as the follow. To simplify the implementation and reduce the call to r.Client.Get()
+				// on kappConfig CR, use the util function to construct the secret name.
+				// TODO: Follow up with reviewer to see if this is feasible
+				Name: util.GenerateDataValueSecretName(cluster.Name, constants.KappControllerAddonName)},
+			},
+		}
+
+		return nil
+	}
+
+	_, err := controllerutil.CreateOrPatch(r.context, r.Client, pkgi, pkgiMutateFn)
+	if err != nil {
+		r.Log.Error(err, fmt.Sprintf("unable to create or patch %s PackageInstall resource for cluster: %s",
+			constants.KappControllerAddonName, cluster.Name))
+		return err
+	}
+
+	return nil
+}
+
+// createOrPatchPackageOnRemote creates the Package CR on remote cluster. In order to install a package on remote cluster
+// the Package CR needs to be present.
+// createOrPatchPackageOnRemote returns a tuple: (<remote-package>, <error>)
+func (r *ClusterBootstrapReconciler) createOrPatchPackageOnRemote(cluster *clusterapiv1beta1.Cluster,
+	cbPkg *runtanzuv1alpha3.ClusterBootstrapPackage, clusterClient client.Client) (*kapppkgv1alpha1.Package, error) {
+
+	var err error
+	// Create or patch Package CR on remote cluster
```

Comment:
Question: Looks like we are not using remote App to deploy kapp controller any more. Instead, we are pushing package and pkgi to remote cluster, while not using pkgr?

Reply:
correct.